<a href="https://colab.research.google.com/github/sivaparvathi1234/siva-parvathi_INFO5502_Spring2022/blob/main/lab_assignment_03_(4).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## The third Lab-assignment (02/10/2022, 50 points in total)

The purpose of this exercise is to understand users' information needs, then collect data from different sources for analysis.

Question 1 (10 points). Fomulate your domain problem: Describe an interesting research question (or practical question) you have in mind, what kind of data should be collected to answer the question(s)? How many data needed for the analysis? The detail steps for collecting and save the data. 

In [ ]:
# Your answer here (no code for this question, write down your answer as detail as possible for the above questions):

'''

Domain Problem: 
. My domain problem is to get the prices of a product from shopping website.

Research question:
. to get the prices in the product, did a research on the vary of prices of that particular product.


Data Required to answer the questions:
The following are the attributes of the required data to answer the research question
. product details
. product with the class of particular attribute.


 i took the 1000 samples from the shopping website  the following link: "https://www.flipkart.com/mens-footwear/sports-shoes/pr?sid=osp,cil,1cu&otracker=nmenu_sub_Men_0_Sports%20Shoes"


Steps for collecting and saving the data:
. We need to use web scraping technique to get the data from that particular website.
. imported BeautifulSoup under bs4 package is a library used to parse HTML & XML
. There are 23 pages of products.
. We parse through each page and go to the products price details to get the prices of that particular product.
. We will get the price from the product homepage.
. We parse the Segment Body-> Paragraph elements in the html code and then store the text data in a json format.
.


'''

Question 2 (10 points). Collect your data to answer the research problem: Write python code to collect 1000 data samples you discussed above.

In [ ]:
#  importing BeautifulSoup under bs4 package is a library used to parse HTML & XML
from bs4 import BeautifulSoup
# urllib module is the URL handling module for python
import urllib
#It helps to define functions and classes, these  help in opening URLs
from urllib.request import Request, urlopen
from urllib.error import HTTPError
# to get the file in json format
import json
import re

total_count = 0
count = 0

prices = []
main_url = "https://www.flipkart.com/mens-footwear/sports-shoes/pr?sid=osp,cil,1cu&otracker=nmenu_sub_Men_0_Sports%20Shoes"
for page_num in range(1, 100):
    # we create the data soup of the main url page html elements.
    link1 = Request(main_url.format(page_num), headers={'User-Agent': 'Mozilla/5.0'})
    url1 = urllib.request.urlopen(link1)
    data1 = url1.read()
    data1_soup = BeautifulSoup(data1)

    print("// Page {} //".format(page_num))
    # we will iterate through pages of the  search results till we get 1000 records
    for narrator_link in data1_soup.find_all('div', attrs = {'class': "_30jeq3"}):
        prices.append(narrator_link.text)
        count = count + 1
        if count == 1000:
            break          
    print(count)
    if count == 1000:
        break
    
print(prices)

// Page 1 //
45
// Page 2 //
90
// Page 3 //
135
// Page 4 //
180
// Page 5 //
225
// Page 6 //
270
// Page 7 //
315
// Page 8 //
360
// Page 9 //
405
// Page 10 //
450
// Page 11 //
495
// Page 12 //
540
// Page 13 //
585
// Page 14 //
630
// Page 15 //
675
// Page 16 //
720
// Page 17 //
765
// Page 18 //
810
// Page 19 //
855
// Page 20 //
900
// Page 21 //
945
// Page 22 //
990
// Page 23 //
1000
['₹1,529', '₹3,959', '₹249', '₹219', '₹499', '₹471', '₹449', '₹499', '₹499', '₹1,047', '₹1,149', '₹398', '₹299', '₹449', '₹598', '₹259', '₹499', '₹1,611', '₹249', '₹299', '₹999', '₹1,049', '₹399', '₹259', '₹584', '₹241', '₹699', '₹1,449', '₹598', '₹574', '₹299', '₹749', '₹249', '₹359', '₹749', '₹259', '₹1,449', '₹749', '₹259', '₹449', '₹259', '₹299', '₹1,611', '₹598', '₹259', '₹1,529', '₹3,959', '₹249', '₹219', '₹499', '₹471', '₹449', '₹499', '₹499', '₹1,047', '₹1,149', '₹398', '₹299', '₹449', '₹598', '₹259', '₹1,049', '₹499', '₹249', '₹299', '₹1,611', '₹999', '₹399', '₹259', '₹584', '₹241

Question 3 (10 points). Understand the data quality: Search a second hand dataset (any dataset) from kaggle or other websites. Describe the data quality problem of the dataset and explain your strtegy to clean the data.

In [ ]:
# Your answer here (no code for this question, write down your answer as detail as possible for the above questions):
'''

Please write you answer here:

Dataset:
. Link: https://www.kaggle.com/shantanudhakadd/email-spam-detection-dataset-classification
. There are 4 data columns in the dataset.
   1) first column is to check whether the email is spam or not spam.
   2)second column is to check the content of the mail.
. The target column  has  2 classes  like Spam or Not Spam and content of the mail.

 In the dataset, data from the email is classified to 5 categeries like valid, mismatched,missing,unique, most common. Based on the percentage, we are able to check whether the email is spam or not.

Strategies to clean the data :
1. we will check the most common emails and start segrigating those mials first.
2.we have to start brushing the active lists from the data.
3. Remove or start merging the duplicates from your mails.
4. we have to remove the spam email addresses.
5.check the missing fields
6. check the duplicate data 
7. check the  invalid formatting
using deep learning model , we can check whether the email is spam or not spam.
'''

Question 4 (20 points). Data cleaning: There are two datasets TwADR-L (from Twitter) and AskAPatient (Link: https://zenodo.org/record/55013#.YgU2NN-ZO4T) for medical concept
normalization. However, the two datasets have serious data quality problems. Please read the introduction of the datasets and clean the two datasets by following the steps in the statement.

In the original dataset, the TwADR-L had 48,057 training, 1,256 validation and 1,427 test examples. The test set (all
test samples from 10 folds combined) consists of 765 unique phrases and 273 unique classes (medical concepts). The AskAPatient dataset contained 156,652 training, 7,926 validation, and 8,662 test examples. The entire test set (all test samples
from 10 folds combined) consists of 3,749 unique phrases and 1,035 unique classes (medical concepts). The authors
randomly split each dataset into ten equal folds, ran 10-fold cross validation and reported the accuracy averaged across the
ten folds. 

We found that, in the original data set, many phrase-label pairs appeared multiple times within the same training data file
and also across the training and test data sets in the same fold. In the AskAPatient data set, on average 35.82% of the test data overlapped with training data in the same fold. In the Twitter (TwADR-L) dataset, on average 8.62% of the test set had an overlap with the training data in the same fold. Having a large overlap between the training and the test data can potentially
introduce bias in the model and contribute to high accuracy. It is not unlikely that the high model performance reported in the original paper may be triggered by the the large overlap between the training and test sets.

Therefore to remove the bias, we further cleaned and recreated the training, validation, and test sets such that each
phrase-label pair appears only once in the entire dataset (either in training, validation or test set).

(1) First, we combined all examples in training, validation and test data from the original data set and then removed all
duplicate phrase-label pairs (examples that have the same phrase and label pair and appear more than once in training/validation/test datasets). Table II shows the statistics of the new dataset (after removing duplicates). The Twitter data set had 3,157 unique phrase-label pairs and 2,220 unique labels (medical concepts) while 173 phrases had multiple labels (i.e., they were assigned to more than one label). Many concepts had only one example, and the concept that had the most number
of examples had 36 phrases. On average, each concept had 1.42 examples. The AskAPatient data set had 4,496 unique phrase-label pairs, 1,036 unique labels while 26 phrases had multiple labels. Table III shows examples of phrases that had multiple labels. For example, ‘mad’ can be mapped to ‘anger’ or ‘rage’ and ‘sore’ can be mapped to ‘pain’ or ‘myalgia’.

(2) Second, we remove all concepts that had less than five examples. The statistics of the final data are shown in Table IV.

(3) Third, we divide all examples without multiple labels into random 10 folds such that each unique phrase-label pair
appears once in one of the 10 test sets. We add the pairs with multiple labels into the training data. This final 10-folds
dataset is used in all our experiments.

(The original paper can be download on canvas)

In [46]:
# You code here (Please add comments in the code):
# 1)
import glob
import requests
from bs4 import BeautifulSoup
import numpy as np
import time
import pandas as pd
import csv
!unzip /content/sample_data/datasets.zip
result_df =  pd.DataFrame()
for dataset in ['TwADR-L', 'AskAPatient']:
    result = {}
    df = pd.DataFrame()
    for filepath in glob.glob('/content/datasets/{}/*.txt'.format(dataset)):
        data = pd.read_csv(filepath, sep = "\t", header = None, encoding= 'unicode_escape')
        df = df.append(data)
        
    df = df.reset_index(drop=True)
    df['phrase_label'] = df[1] + " " + df[2]
    df.columns = ['id', 'labels', 'phrases', 'phrase-label']
    df = df.astype({"id": str})
       
    for columns in df.columns:
        df[columns] = df[columns].str.lower() 

    # dropping duplicates
    df = df.drop_duplicates('phrase-label')

    # storing results of the table in dictionary
    result['Unique_phrases'] = str(len(df['phrases'].unique()))
    result['Unique_labels'] = str(len(df['labels'].unique()))
    result['Unique_phrase_label_pairs'] = str(df.shape[0])
    df1 = pd.DataFrame(df['phrases'].value_counts())
    result['Phrases with multiple labels'] = str(df1[df1['phrases'] > 1].shape[0])
    result['Min examples per label'] = str(df['labels'].value_counts().values.min())
    result['Max examples per label'] = str(df['labels'].value_counts().values.max())
    result['Avg examples per label'] = round(df['labels'].value_counts().mean(),2)
    result_df = result_df.append(result, ignore_index=True)


result_df = result_df.T
# result_df.iloc[0:6].astype(int)
    
result_df.columns = ['TwADR-L', 'AskAPatient']

result_df
   






   
   

Archive:  /content/sample_data/datasets.zip
replace datasets/.DS_Store? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

TwADR-L AskAPatient
Unique_phrases                  2944        4470
Unique_labels                   2220        1038
Unique_phrase_label_pairs       3157        4507
Phrases with multiple labels     173          35
Min examples per label             1           1
Max examples per label            36         141
Avg examples per label          1.42        4.34

In [44]:
import glob
import requests
from bs4 import BeautifulSoup
import numpy as np
import time
import pandas as pd
import csv
!unzip /content/sample_data/datasets.zip
result_df = pd.DataFrame()
# we iterate through both datasets 
for dataset in ['TwADR-L', 'AskAPatient']:
    df = pd.DataFrame()
    result = {}
    for filepath in glob.glob('/content/datasets/{}/*.txt'.format(dataset)):
        #print(filepath)
        data = pd.read_csv(filepath, sep = "\t", header = None, encoding= 'unicode_escape')
        df = df.append(data)
    # we create combined dataframe by combining Each dataset text files.
    df = df.reset_index(drop=True)
    df['phrase_label'] = df[1] + " " + df[2]
    df.columns = ['id', 'labels', 'phrases', 'phrase-label']
    #print(df.shape)
    df = df.astype({"id": str})
    # converting to lower case
    for columns in df.columns:
        df[columns] = df[columns].str.lower() 

    # we drop duplicates that are phrase-label pairs... 
    df = df.drop_duplicates('phrase-label')

    index_list = []
    for i in range(df.shape[0]):
        if df['labels'].value_counts()[df.iloc[i]['labels']] < 5:
            index_list.append(i)

    # we drop the labels that have less than count of 5.
    df.drop(df.index[index_list], inplace=True)
    result = {}

    if dataset == 'TwADR-L':
      Twt_df = df
    else:
      Ask_df = df
 
    # storing results of the table in dictionary
    result['Unique_phrases'] = str(len(df['phrases'].unique()))
    result['Unique_labels'] = str(len(df['labels'].unique()))
    result['Unique_phrase_label_pairs'] = str(df.shape[0])
    df1 = pd.DataFrame(df['phrases'].value_counts())
    result['Phrases with multiple labels'] = str(df1[df1['phrases'] > 1].shape[0])
    result['Min examples per label'] = str(df['labels'].value_counts().values.min())
    result['Max examples per label'] = str(df['labels'].value_counts().values.max())
    result['Avg examples per label'] = round(df['labels'].value_counts().mean(), 2)
    # appending dictinary to dataframe
    result_df = result_df.append(result, ignore_index=True)

result_df = result_df.T
result_df.columns = ['TwADR-L', 'AskAPatient']


result_df

TwADR-L AskAPatient
Unique_phrases                   616        2665
Unique_labels                     76         233
Unique_phrase_label_pairs        721        2686
Phrases with multiple labels      87          19
Min examples per label             5           5
Max examples per label            36         141
Avg examples per label          9.49       11.53